sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -442,9 +442,18 @@ kmac_error_t kmac_init(kmac_operation_t operation,
     return err;
   }
 
+  // We need to preserve some bits of CFG register, such as:
+  // entropy_mode, entropy_ready etc. On the other hand, some bits
+  // need to be reset for each invocation.
   uint32_t cfg_reg =
       abs_mmio_read32(kKmacBaseAddr + KMAC_CFG_SHADOWED_REG_OFFSET);
 
+  // Make sure kmac_en and sideload bits of CFG are reset at each invocation
+  // These bits should be set to 1 only if needed by the rest of the code
+  // in this function.
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_KMAC_EN_BIT, 0);
+  cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_SIDELOAD_BIT, 0);
+
   switch (operation) {
```

Comment:
Can this `switch` statement avoided by defining `kmac_operation_t` constants as the actual values? E.g.
```
// in kmac.h:
typedef enum kmac_operation {
   kKmacOperationSha3 = 0x30,
   ...
} kmac_operation_t;
```

```
// in kmac.c
#include "kmac_regs.h"
...

static_assert(kKmacOperationSha3 == (KMAC_CFG_SHADOWED_MODE_OFFSET << KMAC_CFG_SHADOWED_MODE_OFFSET), "Mismatch);
```

Reply:
If I understood correctly, the idea is to remove switch-case block by doing the following (plus adding static assert checks to ensure that `kmac_operation_t` is in sync with auto-generated header files): 
```
// in kmac.h:
typedef enum kmac_operation {
   kKmacOperationSha3 = 0x0,
   kKmacOperationSHAKE = 0x2,
   kKmacOperationCSHAKE = 0x3,
   kKmacOperationKMAC = 0x4,
} kmac_operation_t;
```
```
// in kmac.c
  // Remove switch-case and do the following
  cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, operation);
```

Unfortunately this won't work for KMAC algorithm. From HW perspective, SHA3, SHAKE, CSHAKE are modes, but KMAC is also cSHAKE mode with additional `kmac_en` bit configured to 1. From SW perspective, each are different operaitons.

Reply:
I see, but you can define the constants as bitfields consisting of a `mode` and `kmac_en` bit.

Reply:
(I think) I see, did I get the idea correctly @alphan:
```
// in kmac.h:
typedef enum kmac_operation {
   kKmacOperationSha3 = 0x00, // SHA-3
   kKmacOperationSHAKE = 0x20, // SHAKE
   kKmacOperationCSHAKE = 0x30, // CSHAKE
   kKmacOperationKMAC = 0x31, // KMAC
} kmac_operation_t;

// in kmac.c:
OT_ASSERT_ENUM_VALUE(kKmacOperationSha3, KMAC_CFG_SHADOWED_MODE_VALUE_SHA3 << KMAC_CFG_SHADOWED_MODE_OFFSET);
OT_ASSERT_ENUM_VALUE(kKmacOperationSHAKE, KMAC_CFG_SHADOWED_MODE_VALUE_SHAKE << KMAC_CFG_SHADOWED_MODE_OFFSET);
OT_ASSERT_ENUM_VALUE(kKmacOperationCSHAKE, KMAC_CFG_SHADOWED_MODE_VALUE_CSHAKE << KMAC_CFG_SHADOWED_MODE_OFFSET);
OT_ASSERT_ENUM_VALUE(kKmacOperationKMAC, (KMAC_CFG_SHADOWED_MODE_VALUE_CSHAKE << KMAC_CFG_SHADOWED_MODE_OFFSET) || (1 << KMAC_CFG_SHADOWED_KMAC_EN_BIT));

// instead of switch-case:
cfg_reg = bitfield_bit32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, operation >> KMAC_CFG_SHADOWED_KMAC_EN_BIT);
cfg_reg = bitfield_field32_write(cfg_reg, KMAC_CFG_SHADOWED_MODE_FIELD, operation >> KMAC_CFG_SHADOWED_MODE_OFFSET);
```

Reply:
I think so but I didn't get the chance to check the values and the `assert`s for correctness.

It would be great if you could follow the pattern [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/lib/drivers/flash_ctrl.h;drc=9d94a804b27459c8af9e8ef0983fb45846706030;l=18): In addition to the constants, you can define macros for bit indices and `bitfield_field32_t`s that can be used to access individual items. Also a comment that illustrates the fields would be helpful, e.g.:

```
- Bit 0: kmac_en
- Bit 1-x: mode
```

Reply:
Thanks for the example @alphan. I force pushed it again with this change. I didn't define macro for `static_assert`s because one of the field is a bit and the other is a field of 2 bits.

This got me thinking though, do we also need to check invalid cases of enum types? For example, is it possible to see something like `operation = 0x55` which is undefined enum (which would have gone to invalid case in switch-case)? AFAUI, we will have some hardened enum types at some point, so are we also concerned about faults on the bits of enum-typed variables?

Reply:
We need to consider this case-by-case, e.g. we can't gain anything useful from this check if the bits we care about are fully mapped (0/1: `kmac_en`). For a switch-case, we definitely need to do this + use hardened enums.